"install_timeout" configuration option don't used for creating environment using conda

At first glance, it seems to me that this option should affect the `timeout` parameter of `check_output` function which is used to run conda.

https://github.com/airspeed-velocity/asv/blob/6d4d845bd3a845555f982da0a45d935eca550857/asv/util.py#L315

The error:
```bash
·· Failure creating environment for conda-py3.7
    result = args.func(args)
             ^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/commands/__init__.py", line 49, in run_from_args
    return cls.run_from_conf_args(conf, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/commands/run.py", line 183, in run_from_conf_args
    return cls.run(
           ^^^^^^^^
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/commands/run.py", line 294, in run
    Setup.perform_setup(environments, parallel=parallel)
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/commands/setup.py", line 89, in perform_setup
    list(map(_create, environments))
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/commands/setup.py", line 21, in _create
    env.create()
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/environment.py", line 704, in create
    self._setup()
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/plugins/conda.py", line 174, in _setup
    self._run_conda(['env', 'create', '-f', env_file_name,
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/plugins/conda.py", line 227, in _run_conda
    return util.check_output([conda] + args, env=env)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/lib/python3.11/site-packages/asv/util.py", line 754, in check_output
    raise ProcessError(args, retcode, stdout, stderr)
asv.util.ProcessError: Command '/usr/share/miniconda/bin/conda env create -f ../requirements/env_hdk.yml -p /home/runner/work/modin/modin/asv_bench/.asv/env/6df994238d3562d4e97431b9f4cec556 --force' timed out
Error: Process completed with exit code 1.
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -2,6 +2,10 @@ name: CI
 
 on: [push, pull_request]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
   test:
     name: test

--- a/.github/workflows/pre-commit.yml
+++ b/.github/workflows/pre-commit.yml
@@ -5,6 +5,10 @@ on:
   push:
     branches: [master]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
   pre-commit:
     runs-on: ubuntu-latest

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -17,6 +17,8 @@ API Changes
 
 Bug Fixes
 ^^^^^^^^^
+- Fixed ``install_timeout`` for ``conda`` (#1310)
+- Fixed handling of local ``pip`` matrix (#1312)
 - Fixed the deadlock when mamba is used with an environment file. (#1300)
 - Fixed environment file usage with mamba and recognizes default
   ``environment.yml``. (#1303)

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -8,6 +8,7 @@ New Features
 - Benchmarks can be skipped during their execution (after setup) by raising ``SkipNotImplemented`` (#1307)
 - Added ``default_benchmark_timeout`` to the configuration object, can also be
   passed via ``-a timeout=NUMBER`` (#1308)
+- ``ASV_RUNNER_PATH`` can be set from the terminal to test newer versions of ``asv_runner`` (#1312)
 
 API Changes
 ^^^^^^^^^^^

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -521,8 +521,9 @@ class Environment:
 
         # Update the _base_requirements if needed
         for key in list(self._requirements.keys()):
-            self._base_requirements[key] = self._requirements[key]
-            del self._requirements[key]
+            if key in self._base_requirements:
+                self._base_requirements[key] = self._requirements[key]
+                del self._requirements[key]
 
         self._build_command = conf.build_command
         self._install_command = conf.install_command

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -508,7 +508,8 @@ class Environment:
         # These are needed for asv to build and run the project, not part of
         # benchmark name mangling
         self._base_requirements = {}
-        self._base_requirements["pip+asv_runner"] = ""
+        # gh-1314
+        self._base_requirements["pip+asv_runner"] = os.getenv("ASV_RUNNER_PATH", "")
         if not util.ON_PYPY:
             # XXX: What if pypy installed asv tries to benchmark a cpython
             # python?

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -519,6 +519,11 @@ class Environment:
         if (Path.cwd() / "pdm.lock").exists():
             self._base_requirements["pdm"] = ""
 
+        # Update the _base_requirements if needed
+        for key in list(self._requirements.keys()):
+            self._base_requirements[key] = self._requirements[key]
+            del self._requirements[key]
+
         self._build_command = conf.build_command
         self._install_command = conf.install_command
         self._uninstall_command = conf.uninstall_command

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -975,6 +975,7 @@ class Environment:
                              PIP_USER=str("false"),
                              PATH=str(os.pathsep.join(paths)))
         exe = self.find_executable(executable)
+        kwargs["timeout"] = self._install_timeout
         return util.check_output([exe] + args, **kwargs)
 
     def load_info_file(self, path):

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -12,6 +12,7 @@ import re
 import sys
 import itertools
 import subprocess
+import importlib
 from pathlib import Path
 
 from .console import log
@@ -509,7 +510,26 @@ class Environment:
         # benchmark name mangling
         self._base_requirements = {}
         # gh-1314
-        self._base_requirements["pip+asv_runner"] = os.getenv("ASV_RUNNER_PATH", "")
+        asv_runner_path = os.getenv("ASV_RUNNER_PATH", "")
+        module_path = Path(asv_runner_path) / "asv_runner"
+
+        # Check if the path points to a directory containing the "asv_runner" module
+        if module_path.is_dir() and (module_path / "__init__.py").is_file():
+            spec = importlib.util.spec_from_file_location("asv_runner",
+                                                          module_path / "__init__.py")
+            # Attempt to load the module
+            asv_runner_module = importlib.util.module_from_spec(spec)
+            try:
+                spec.loader.exec_module(asv_runner_module)
+                self._base_requirements["pip+asv_runner"] = asv_runner_path
+            except Exception as e:
+                self._base_requirements["pip+asv_runner"] = ""
+                log.warning(f"Failed to load module from ASV_RUNNER_PATH: {e}")
+        else:
+            self._base_requirements["pip+asv_runner"] = ""
+            if asv_runner_path:
+                log.warning("ASV_RUNNER_PATH does not point"
+                            "to a directory containing the 'asv_runner' module")
         if not util.ON_PYPY:
             # XXX: What if pypy installed asv tries to benchmark a cpython
             # python?

--- a/asv/environment.py
+++ b/asv/environment.py
@@ -976,7 +976,8 @@ class Environment:
                              PIP_USER=str("false"),
                              PATH=str(os.pathsep.join(paths)))
         exe = self.find_executable(executable)
-        kwargs["timeout"] = self._install_timeout
+        if kwargs.get("timeout", None) is None:
+            kwargs["timeout"] = self._install_timeout
         return util.check_output([exe] + args, **kwargs)
 
     def load_info_file(self, path):

--- a/asv/plugins/conda.py
+++ b/asv/plugins/conda.py
@@ -87,6 +87,8 @@ class Conda(environment.Environment):
         self._python = python
         self._requirements = requirements
         self._conda_channels = conf.conda_channels
+        if "conda-forge" not in conf.conda_channels:
+            self._conda_channels += ["conda-forge"]
         self._conda_environment_file = conf.conda_environment_file
 
         if conf.conda_environment_file == "IGNORE":
@@ -143,13 +145,8 @@ class Conda(environment.Environment):
             # categorize & write dependencies based on pip vs. conda
             env_file.writelines((f'   - {s}\n' for s in conda_args))
             if pip_args:
-                # and now specify the packages that are to be installed in
-                # the pip subsection
                 env_file.write('   - pip:\n')
-                env_file.writelines((f'     - {s}\n' for s in pip_args))
-
             env_file.close()
-
             try:
                 env_file_name = self._conda_environment_file or env_file.name
                 self._run_conda(['env', 'create', '-f', env_file_name,
@@ -174,6 +171,12 @@ class Conda(environment.Environment):
                 raise
         finally:
             os.unlink(env_file.name)
+        if not len(pip_args) == 0:
+            pip_calls = []
+            for pkgname, pipval in pip_args:
+                pip_calls.append(util.construct_pip_call(self._run_pip, pkgname, pipval))
+            for pipcall in pip_calls:
+                pipcall()
 
     def _get_requirements(self):
         conda_args = []
@@ -181,11 +184,11 @@ class Conda(environment.Environment):
 
         for key, val in {**self._requirements,
                          **self._base_requirements}.items():
-            if key.startswith('pip+'):
+            if key.startswith("pip+"):
                 if val:
-                    pip_args.append(f"{key[4:]}=={val}")
+                    pip_args.append((key[4:], val))
                 else:
-                    pip_args.append(key[4:])
+                    pip_args.append((key[4:], None))
             else:
                 if val:
                     conda_args.append(f"{key}={val}")
@@ -224,3 +227,8 @@ class Conda(environment.Environment):
 
         with lock():
             return super(Conda, self).run_executable(executable, args, **kwargs)
+
+    def _run_pip(self, args, **kwargs):
+        # Run pip via python -m pip, so that it works on Windows when
+        # upgrading pip itself, and avoids shebang length limit on Linux
+        return self.run_executable("python", ["-mpip"] + list(args), **kwargs)

--- a/asv/plugins/conda.py
+++ b/asv/plugins/conda.py
@@ -207,7 +207,9 @@ class Conda(environment.Environment):
             raise util.UserError(str(e))
 
         with _conda_lock():
-            return util.check_output([conda] + args, env=env)
+            return util.check_output([conda] + args,
+                                     timeout=self._install_timeout,
+                                     env=env)
 
     def run(self, args, **kwargs):
         log.debug(f"Running '{' '.join(args)}' in {self.name}")

--- a/asv/plugins/mamba.py
+++ b/asv/plugins/mamba.py
@@ -122,8 +122,11 @@ class Mamba(environment.Environment):
             transaction = solver.solve(mamba_pkgs)
             transaction.execute(libmambapy.PrefixData(self._path))
         if not len(pip_args) == 0:
-            pargs = ["install", "-v", "--upgrade-strategy", "only-if-needed"]
-            self._run_pip(pargs + pip_args)
+            pip_calls = []
+            for pkgname, pipval in pip_args:
+                pip_calls.append(util.construct_pip_call(self._run_pip, pkgname, pipval))
+            for pipcall in pip_calls:
+                pipcall()
 
     def _get_requirements(self):
         mamba_args = []
@@ -133,9 +136,9 @@ class Mamba(environment.Environment):
                          **self._base_requirements}.items():
             if key.startswith("pip+"):
                 if val:
-                    pip_args.append(f"{key[4:]}=={val}")
+                    pip_args.append((key[4:], val))
                 else:
-                    pip_args.append(key[4:])
+                    pip_args.append((key[4:], None))
             else:
                 if val:
                     mamba_args.append(f"{key}={val}")

--- a/asv/plugins/mamba.py
+++ b/asv/plugins/mamba.py
@@ -83,7 +83,10 @@ class Mamba(environment.Environment):
         if not re.match(r'^[0-9].*$', python):
             return False
         else:
-            mamba_path = str(Path(os.getenv("CONDA_EXE")).parent / "mamba")
+            if os.getenv("CONDA_EXE"):
+                mamba_path = str(Path(os.getenv("CONDA_EXE")).parent / "mamba")
+            else:
+                return False
             try:
                 return util.search_channels(mamba_path, "python", python)
             except util.ProcessError:

--- a/asv/plugins/virtualenv.py
+++ b/asv/plugins/virtualenv.py
@@ -2,6 +2,7 @@
 import sys
 import re
 import os
+from pathlib import Path
 
 from packaging.version import Version
 
@@ -156,7 +157,10 @@ class Virtualenv(environment.Environment):
                 pkg = key[4:]
 
             if val:
-                args.append(f"{pkg}=={val}")
+                if Path(val).is_dir():
+                    args.append(val)
+                else:
+                    args.append(f"{pkg}=={val}")
             else:
                 args.append(pkg)
         self._run_pip(args, timeout=self._install_timeout, env=env)

--- a/asv/plugins/virtualenv.py
+++ b/asv/plugins/virtualenv.py
@@ -148,14 +148,18 @@ class Virtualenv(environment.Environment):
 
         self._run_pip(pip_args, env=env)
         pip_calls = []
+        pip_args = []
 
         for key, val in {**self._requirements,
                          **self._base_requirements}.items():
-            pkg = key
-            if key.startswith('pip+'):
-                pkg = key[4:]
-            pip_calls.append(util.construct_pip_call(self._run_pip, pkg, val))
+            if key.startswith("pip+"):
+                if val:
+                    pip_args.append((key[4:], val))
+                else:
+                    pip_args.append((key[4:], None))
 
+        for pkgname, pipval in pip_args:
+            pip_calls.append(util.construct_pip_call(self._run_pip, pkgname, pipval))
         for pipcall in pip_calls:
             pipcall()
 

--- a/asv/plugins/virtualenv.py
+++ b/asv/plugins/virtualenv.py
@@ -147,12 +147,11 @@ class Virtualenv(environment.Environment):
         env.update(self.build_env_vars)
 
         self._run_pip(pip_args, env=env)
-        pip_args = ['install', '-v', '--upgrade']
+        _args = ['install', '-v', '--upgrade']
         pcalls = util.construct_pip_calls(self._run_pip, {**self._requirements,
-                                                         **self._base_requirements}, pip_args)
+                                                         **self._base_requirements}, _args)
         for pcall in pcalls:
-            pcall()
-        # self._run_pip(args, timeout=self._install_timeout, env=env)
+            pcall(timeout=self._install_timeout, env=env)
 
     def _run_pip(self, args, **kwargs):
         # Run pip via python -m pip, so that it works on Windows when

--- a/asv/plugins/virtualenv.py
+++ b/asv/plugins/virtualenv.py
@@ -147,11 +147,17 @@ class Virtualenv(environment.Environment):
         env.update(self.build_env_vars)
 
         self._run_pip(pip_args, env=env)
-        _args = ['install', '-v', '--upgrade']
-        pcalls = util.construct_pip_calls(self._run_pip, {**self._requirements,
-                                                         **self._base_requirements}, _args)
-        for pcall in pcalls:
-            pcall(timeout=self._install_timeout, env=env)
+        pip_calls = []
+
+        for key, val in {**self._requirements,
+                         **self._base_requirements}.items():
+            pkg = key
+            if key.startswith('pip+'):
+                pkg = key[4:]
+            pip_calls.append(util.construct_pip_call(self._run_pip, pkg, val))
+
+        for pipcall in pip_calls:
+            pipcall()
 
     def _run_pip(self, args, **kwargs):
         # Run pip via python -m pip, so that it works on Windows when

--- a/asv/plugins/virtualenv.py
+++ b/asv/plugins/virtualenv.py
@@ -2,7 +2,6 @@
 import sys
 import re
 import os
-from pathlib import Path
 
 from packaging.version import Version
 
@@ -148,22 +147,12 @@ class Virtualenv(environment.Environment):
         env.update(self.build_env_vars)
 
         self._run_pip(pip_args, env=env)
-
-        args = ['install', '-v', '--upgrade']
-        for key, val in {**self._requirements,
-                         **self._base_requirements}.items():
-            pkg = key
-            if key.startswith('pip+'):
-                pkg = key[4:]
-
-            if val:
-                if Path(val).is_dir():
-                    args.append(val)
-                else:
-                    args.append(f"{pkg}=={val}")
-            else:
-                args.append(pkg)
-        self._run_pip(args, timeout=self._install_timeout, env=env)
+        pip_args = ['install', '-v', '--upgrade']
+        pcalls = util.construct_pip_calls(self._run_pip, {**self._requirements,
+                                                         **self._base_requirements}, pip_args)
+        for pcall in pcalls:
+            pcall()
+        # self._run_pip(args, timeout=self._install_timeout, env=env)
 
     def _run_pip(self, args, **kwargs):
         # Run pip via python -m pip, so that it works on Windows when

--- a/asv/util.py
+++ b/asv/util.py
@@ -1304,13 +1304,11 @@ def construct_pip_call(pip_caller, pkgname, pipval=None):
     if pipval:
         if Path(pipval).is_dir():
             pargs += [pipval]
-            return functools.partial(pip_caller, pargs)
         else:
             pargs += [f"{pkgname}=={pipval}"]
-            return functools.partial(pip_caller, pargs)
     else:
         pargs += [pkgname]
-        return functools.partial(pip_caller, pargs)
+    return functools.partial(pip_caller, pargs)
 
 
 if hasattr(sys, 'pypy_version_info'):

--- a/asv/util.py
+++ b/asv/util.py
@@ -1302,12 +1302,12 @@ def search_channels(cli_path, pkg, version):
 def construct_pip_call(pip_caller, pkgname, pipval=None):
     pargs = ['install', '-v', '--upgrade']
     if pipval:
-        if ('-' or '--' in pipval):
-            ptokens = pipval.split()
-            pipval = [x for x in ptokens if Path(x).is_dir()][0]
-            pargs += [x for x in ptokens if '-' in x]
-        if Path(pipval).is_dir():
-            pargs += [pipval]
+        ptokens = pipval.split()
+        flags = [x for x in ptokens if x.startswith('-')]
+        paths = [x for x in ptokens if Path(x).is_dir()]
+        pargs += flags
+        if paths:
+            pargs += paths
         else:
             pargs += [f"{pkgname}=={pipval}"]
     else:

--- a/asv/util.py
+++ b/asv/util.py
@@ -1299,7 +1299,7 @@ def search_channels(cli_path, pkg, version):
     return True
 
 
-def construct_pip_call(pip_caller, pkgname, pipval):
+def construct_pip_call(pip_caller, pkgname, pipval=None):
     pargs = ['install', '-v', '--upgrade']
     if pipval:
         if Path(pipval).is_dir():

--- a/asv/util.py
+++ b/asv/util.py
@@ -1299,24 +1299,18 @@ def search_channels(cli_path, pkg, version):
     return True
 
 
-def construct_pip_calls(pip_caller, pip_reqs, args, **kwargs):
-    fcalls = []
-    for key, val in pip_reqs.items():
-        pargs = ['install', '-v']
-        pkg = key
-        if key.startswith('pip+'):
-            pkg = key[4:]
-        if val:
-            if Path(val).is_dir():
-                pargs.append(val)
-                fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
-            else:
-                pargs.extend(['--upgrade', f"{pkg}=={val}"])
-                fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
+def construct_pip_call(pip_caller, pkgname, pipval):
+    pargs = ['install', '-v', '--upgrade']
+    if pipval:
+        if Path(pipval).is_dir():
+            pargs += [pipval]
+            return functools.partial(pip_caller, pargs)
         else:
-            pargs.append(pkg)
-            fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
-    return fcalls
+            pargs += [f"{pkgname}=={pipval}"]
+            return functools.partial(pip_caller, pargs)
+    else:
+        pargs += [pkgname]
+        return functools.partial(pip_caller, pargs)
 
 
 if hasattr(sys, 'pypy_version_info'):

--- a/asv/util.py
+++ b/asv/util.py
@@ -23,6 +23,8 @@ import shlex
 import operator
 import collections
 import multiprocessing
+import functools
+from pathlib import Path
 
 import json5
 from asv_runner.util import human_time, human_float, _human_time_units
@@ -1295,6 +1297,26 @@ def search_channels(cli_path, pkg, version):
         return False
     # Worked!
     return True
+
+
+def construct_pip_calls(pip_caller, pip_reqs, args, **kwargs):
+    fcalls = []
+    for key, val in pip_reqs.items():
+        pargs = ['install', '-v']
+        pkg = key
+        if key.startswith('pip+'):
+            pkg = key[4:]
+        if val:
+            if Path(val).is_dir():
+                pargs.append(val)
+                fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
+            else:
+                pargs.extend(['--upgrade', f"{pkg}=={val}"])
+                fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
+        else:
+            pargs.append(pkg)
+            fcalls.append(functools.partial(pip_caller, pargs, **kwargs))
+    return fcalls
 
 
 if hasattr(sys, 'pypy_version_info'):

--- a/asv/util.py
+++ b/asv/util.py
@@ -1302,6 +1302,10 @@ def search_channels(cli_path, pkg, version):
 def construct_pip_call(pip_caller, pkgname, pipval=None):
     pargs = ['install', '-v', '--upgrade']
     if pipval:
+        if ('-' or '--' in pipval):
+            ptokens = pipval.split()
+            pipval = [x for x in ptokens if Path(x).is_dir()][0]
+            pargs += [x for x in ptokens if '-' in x]
         if Path(pipval).is_dir():
             pargs += [pipval]
         else:

--- a/docs/source/asv.conf.json.rst
+++ b/docs/source/asv.conf.json.rst
@@ -267,6 +267,11 @@ the project being benchmarked may specify in its ``setup.py`` file.
     name with ``pip+``. For example, ``emcee`` is only available from ``pip``,
     so the package name to be used is ``pip+emcee``.
 
+    .. versionadded::0.6
+
+      ``pip`` dependencies can now accept local (fully qualified) directories,
+      and also take flags (e.g. ``-e``)
+
 The ``env`` and ``env_nobuild`` dictionaries can be used to set also
 environment variables::
 

--- a/docs/source/dev.rst
+++ b/docs/source/dev.rst
@@ -5,6 +5,7 @@ This section describes some things that may be of interest to
 developers and other people interested in internals of ``asv``.
 
 .. note::
+
    From version 0.6 on-wards, functionality in ``asv`` has been split into the
    section needed by the code being benchmarked (``asv_runner``) and the rest of
    ``asv``. This means that the ``asv`` documentation covers setting up

--- a/docs/source/env_vars.rst
+++ b/docs/source/env_vars.rst
@@ -25,6 +25,14 @@ behavior are also set:
 - ``PYTHONNOUSERSITE``: ``True`` (for conda environments only)
 - ``PYTHONPATH``: unset (if really needed, can be overridden by setting ``ASV_PYTHONPATH``)
 
+.. note::
+
+    .. versionadded::0.6
+
+      ``ASV_RUNNER_PATH`` may be set to provide a local installation of
+      ``asv_runner``, mostly used for the CI to ensure changes to ``asv_runner``
+      do not break ``asv``
+
 
 Custom environment variables
 ----------------------------

--- a/docs/source/writing_benchmarks.rst
+++ b/docs/source/writing_benchmarks.rst
@@ -187,9 +187,10 @@ value is the maximum of the timeouts of the benchmarks using it.
 
 .. note::
 
-   From version 0.6 onwards, the configuration option
-   ``default_benchmark_timeout`` can also be set for a project-wide
-   timeout.
+    .. versionchanged:: 0.6
+
+        The configuration option ``default_benchmark_timeout``
+        can also be set for a project-wide timeout.
 
 .. _benchmark-attributes:
 

